### ISSUE TITLE
Abort the file upload

### ISSUE THREAD

**Issue by paragshar** [balderdashy/skipper#80]
I am uploading a file along with a few other fields. 
Field shared_with , type: text
Field media, type: file

req.file("media").upload(function(err, files){
});

Before I call the code above,  I check for validity of other fields in my Sails policies. Whenever validation fails, I will like to send back the response to the invoking party telling them about the validation error. I do it in exactly the same way as I send the response in case there is no file being uploaded.

However in case when I am trying to upload a file, 
res.badRequest() does not have any effect as if the statement itself is not getting executed.

I realised that the above scenario happens only when I upload a file. In case I do not upload a file, it works like charm without any problem.

However the above thing works if I try to save the uploaded file like below, and then send back the error from within this code like below

req.file('media').upload({maxBytes: Number.MAX_VALUE}, function(err, files){
        return res.status(400).json({message: "'shared_with' should be a non empty array"});
});

Ideally in case of validation failure, I would like to send the error without really saving the file. 

Any clues? Is there a way I can abort the file upload process, and simply send back the error status code?

**Comment by JonGarbayo** [comment]
Same issue for me :(
If i do a `res.something()` (like `badRequest()`, or even a personal status followed by some JSON data `status(409).json({ error: 'user already exists' })`), without invoking the `req.file('file').upload()` method, the result never arrives to the client. The spinner is turning again and again...

A little difference for me: with a file in the input or not, but while there is the `enctype` attribute, the infinite loop happens.


# My hacking solution
Since you've posted this issue, maybe you found a hack to get around it. But I would like to share mine, for you and maybe other Sails.js developers :)

Uploading the file have to be the first thing you do in your controller. No issue doing this.
And if an error occurs? Just delete the file :o !

I'v created a little nice function to do this. It fits right for my needs, but maybe not for everyone. Feel free to get it and adapt it :)

```javascript
/**
 * Delete a file in app file system. Optionally execute a callback 
 * function once it's done (recommended).
 * 
 * @param  {String}   file_descriptor     File path. Use file.fd to
 * get it.
 * 
 * @param  {Function} callback            Optional. Function to
 * execute right after file deletion.
 * 
 * @param  {Array}    callback_parameters Optional. Parameters to pass
 * to the callback function.
 * 
 * @return {void}
 */
deleteFile: function (file_descriptor, callback, callback_parameters) 
{
	var SkipperDisk = require('skipper-disk');
	var file_adapter = SkipperDisk();

	file_adapter.rm(file_descriptor,
	function whenDone()
	{
		if (callback && typeof(callback) === "function")
		{
			if (callback_parameters)
			{
				callback.apply(this, callback_parameters);
			}
			else
			{
				callback();
			}
		}
	});
} 
```

**Comment by metzuda** [comment]
@paragshar  Why do you need to use policies to check field validity? I'm just curious.

@JonGarbayo uploading a file is NOT the firs thing you have to do in a controller action, unless you need that file to be uploaded before you can do anything else within that action.

Here's what I do, works fine. 

1. grab URL params (e.g. ``` var id = req.param('id'); ```
2. grab req.body for the non-file form data; e.g.  ``` var formData = req.body; ``
3. verify form data, send res.badRequest() if it's invalid; e.g.  ```if(!formData || !formData.length) return res.badRequest();```
4. upload file; e.g. ```req.file('media').upload({uploadParams}, function(e, r){ /* do stuff */});```

**Comment by Tumetsu** [comment]
I'm having similar problem as @paragshar At the moment I can handle invalid request only after upload has completed. Before that returning from controller function does nothing and client will never get response. @JonGarbayo 's solution reminds my own where I rollback from invalid request by deleting uploaded file. 

However, the same problem arises if I have middleware which checks if the user is authenticated before upload. If user is unauthenticated, file upload prevents my server from responding with 403.

I tried @metzuda way but it doesn't work if I try to return response before upload function runs.

**Comment by sailsbot** [comment]
Thanks for posting, @paragshar.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by 1upon0** [comment]
No updates on this?
